sdk/core/Azure.Core/src/Shared/EventSourceEventFormatting.cs

```diff
@@ -1,79 +1,125 @@
+                return string.Format(CultureInfo.InvariantCulture, eventData.Message, payloadArray);
             }
-
-            var stringBuilder = new StringBuilder();
-            stringBuilder.Append(eventData.EventName);
-
-            if (!string.IsNullOrWhiteSpace(eventData.Message))
+            catch (FormatException)
             {
-                stringBuilder.AppendLine();
-                stringBuilder.Append(nameof(eventData.Message)).Append(" = ").Append(eventData.Message);
             }
+        }
 
-            if (eventData.PayloadNames != null)
-            {
-                for (int i = 0; i < eventData.PayloadNames.Count; i++)
-                {
-                    stringBuilder.AppendLine();
-                    stringBuilder.Append(eventData.PayloadNames[i]).Append(" = ").Append(payloadArray[i]);
-                }
-            }
+        StringBuilder stringBuilder = RentStringBuilder();
+        stringBuilder.Append(eventData.EventName);
 
-            return stringBuilder.ToString();
+        if (!string.IsNullOrWhiteSpace(eventData.Message))
+        {
+            stringBuilder.AppendLine();
+            stringBuilder.Append(nameof(eventData.Message)).Append(" = ").Append(eventData.Message);
         }
 
-        private static void ProcessPayloadArray(object?[] payloadArray)
+        if (eventData.PayloadNames != null)
         {
-            for (int i = 0; i < payloadArray.Length; i++)
+            for (int i = 0; i < eventData.PayloadNames.Count; i++)
             {
-                payloadArray[i] = FormatValue(payloadArray[i]);
+                stringBuilder.AppendLine();
+                stringBuilder.Append(eventData.PayloadNames[i]).Append(" = ").Append(payloadArray[i]);
             }
         }
 
-        private static object? FormatValue(object? o)
+        return ToStringAndReturnStringBuilder(stringBuilder);
+    }
+
+    private static void ProcessPayloadArray(object?[] payloadArray)
+    {
+        for (int i = 0; i < payloadArray.Length; i++)
         {
-            if (o is byte[] bytes)
+            payloadArray[i] = FormatValue(payloadArray[i]);
+        }
+    }
+
+    private static object? FormatValue(object? o)
+    {
+        if (o is byte[] bytes)
+        {
+#if NET6_0_OR_GREATER
+            return Convert.ToHexString(bytes);
+#else
+            int length = 2 * bytes.Length; // Two hex characters per byte
+            if (length < 256)
             {
-                var stringBuilder = new StringBuilder();
-                foreach (byte b in bytes)
+                static char ToHex(int value) => (char)(value < 10 ? value + '0' : value - 10 + 'A');
+
+                Span<char> buffer = stackalloc char[length];
+                for (int i = 0; i < bytes.Length; i++)
                 {
-                    stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "{0:X2}", b);
+                    byte b = bytes[i];
+                    buffer[i * 2] = ToHex(b >> 4);
```

Comment:
This isn't going to be obvious to many, including some who may be maintaining this.   Can we please add comments describing the approach used and reason for bit manipulation?

Reply:
Added comments